Login: Add new screen for site credential login in the Jetpack setup flow

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #8075
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a new screen for signing in with site credentials for the Jetpack setup flow. Changes include:
- A new coordinator to handle the native Jetpack setup flow.
- A new view and view model for site credential login. 
- Updated the new Jetpack error screen with the entry point to the native Jetpack setup flow.
- Handled credential login.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Prerequisite: Make sure that you have a test store that is not associated with your WP.com account or doesn't have Jetpack.

- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- If your account doesn't have any associated site, notice that the empty store picker is displayed with the matching design as on Figma.
- Select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- Notice that a new screen for site credential login is presented. The UI should match the design in ERGtCFLyiC9CFpHsw1XIid-fi-1110%3A37319. 
- Enter incorrect credentials, and notice that an alert is displayed with the appropriate message.
- Try again with the correct credentials, notice in Xcode console: `🎉`. The actual implementation will be handled in a subsequent PR.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->
<img src="https://user-images.githubusercontent.com/5533851/202393271-ebab2f44-bbe6-4fa9-81af-e129d7236784.png" width=320 />

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add new view for site credential login

itsmeichigo committed:
Extract JetpackInstallHeaderView for reuse

itsmeichigo committed:
Merge branch 'trunk' into issue/8075-site-credential-screen

itsmeichigo committed:
Add coordinator to handle Jetpack setup during login

itsmeichigo committed:
Update SiteCredentialLoginView and add hosting controller

itsmeichigo committed:
Present site credential screen from the coordinator

itsmeichigo committed:
Start coordinator from jetpack error screen

itsmeichigo committed:
Add bottom button for site credential screen

itsmeichigo committed:
Update layout for site credential login view

itsmeichigo committed:
Add cancel button for SiteCredentialLoginHostingViewController

itsmeichigo committed:
Add view model for SiteCredentialLoginView

itsmeichigo committed:
Add test for SiteCredentialLoginViewModel

itsmeichigo committed:
Dismiss keyboard when tapping primary button

itsmeichigo committed:
Add background color for bottom button

itsmeichigo committed:
Handle site credential login

itsmeichigo committed:
Update unit tests for SiteCredentialLoginViewModel

itsmeichigo committed:
Fix test failure due to uninitialized WordPressAuthenticator

jaclync commented on WooCommerce/Classes/Authentication/Jetpack Setup/LoginJetpackSetupCoordinator.swift:
```diff
@@ -0,0 +1,29 @@
+    private let analytics: Analytics
+
+    init(siteURL: String,
+         connectionOnly: Bool,
```
nit: a comment on this property might be helpful? I wasn't sure what this is until I look down in the chain to `SiteCredentialLoginView`

itsmeichigo replied:
Yup, updated in c64e64b2935ae254c779acdd7125c41124466650 for all the occurrences of `connectionOnly` 😅

jaclync commented on WooCommerce/Classes/Authentication/Jetpack Setup/Site Credential Login/SiteCredentialLoginView.swift:
```diff
@@ -0,0 +1,199 @@
+                keyboardIsShown = false
+                viewModel.handleLogin()
+            } label: {
+                Text(connectionOnly ? Localization.connectJetpack : Localization.installJetpack)
```
I'm probably missing some context here - it looks like this button is to log in for both `connectionOnly` states, then what are the differences between these two states? 🤔

itsmeichigo replied:
I understand from Joe's design that we want to consider this a step of the Jetpack setup flow, not login. We want to avoid giving the impression that the user has to log in twice, first with a WP.com account and then with site credentials.

jaclync replied:
The title says `Log in to your store` and the subtitle also starts with `Log in to ...` 😆

jaclync commented on WooCommerce/Classes/Authentication/Jetpack Setup/Site Credential Login/SiteCredentialLoginView.swift:
```diff
@@ -0,0 +1,199 @@
+                        } else {
+                            SecureField(Localization.enterPassword, text: $viewModel.password)
+                                .focused($keyboardIsShown)
+                        }
```
just wondering, is `SecureField` meant to be under the `showsSecureInput` condition to match its description?

itsmeichigo replied:
Good catch! Updated this in 9cffa3dc303744f39ee2413c700cecad24b329c4.

jaclync commented on WooCommerce/Classes/Authentication/Jetpack Setup/Site Credential Login/SiteCredentialLoginView.swift:
```diff
@@ -0,0 +1,199 @@
+                Spacer()
+            }
+        }
+        .safeAreaInset(edge: .bottom, content: {
```
TIL `safeAreaInset` that is iOS 15+ 👍 will try using this instead of another layer of VStack next time!

jaclync commented on WooCommerce/Classes/Authentication/Jetpack Setup/Site Credential Login/SiteCredentialLoginView.swift:
```diff
@@ -0,0 +1,199 @@
+                        Button(action: {
+                            showsSecureInput.toggle()
+                        }) {
+                            Image(systemName: showsSecureInput ? "eye.slash" : "eye")
```
I wish there was an easier way to reuse this component with the one in account creation, but it looks like the text field style is different

itsmeichigo replied:
Yes I looked into doing that but ended up copying your code 😆

jaclync commented on WooCommerce/Classes/Authentication/Jetpack Setup/Site Credential Login/SiteCredentialLoginViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+        isLoggingIn = false
+
+        let err = error as NSError
+        errorMessage = err.code == 3 ? Localization.wrongCredentials : Localization.genericFailure
```
nit: is there any reference for this error code? feels pretty magic 😅

itsmeichigo replied:
This was incorrect 😅 I updated the check in 78c604c8b79908d42391f94abb1d06261b560b13 to match [the behavior](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/blob/6ca641d4655bacf228073c975639e9ee81827a2a/WordPressAuthenticator/Unified%20Auth/View%20Related/Site%20Address/SiteCredentialsViewController.swift#L515-L525) in `WordPressAuthenticator`.

jaclync reviewed (approved):
Works as expected, excited to see the native Jetpack setup flow! :shipit: 

Minor non-blocking observation on how the bottom content is scrolled through the primary CTA, maybe the CTA can use some top padding?

<img src="https://user-images.githubusercontent.com/1945542/202466280-177a3a32-a5e5-4142-9f35-f0cf2bd5fc4b.png" width="300" />

itsmeichigo committed:
Add comments for `connectionOnly`

itsmeichigo committed:
Use `SecureField` for showsSecureInput

itsmeichigo committed:
Add top padding to bottom button

itsmeichigo committed:
Update error code and domain check for wrong credentials

itsmeichigo committed:
Fix layout for the eye button

itsmeichigo committed:
Add reset button

itsmeichigo commented:
Thank you for the review @jaclync! 

> Minor non-blocking observation on how the bottom content is scrolled through the primary CTA, maybe the CTA can use some top padding?

Good catch - I added the padding in 23365649854ea4378a3879948bccae322fb92916.

I've also fixed an issue with the text on the password field and added the missing Reset password button in c08c7df1e00cad5177476d3e49eacabd718752c1 and 38807469e16d7f6501f56cd8d4faca12253476cc 😅. Please feel free to take another look if you can, thanks!

<img src="https://user-images.githubusercontent.com/5533851/202611302-8086d629-4613-4f12-833c-f925d934d4b1.png" width=320 />

jaclync commented:
Still LGTM 👍 I retried the failing UI tests CI step, hopefully it's green and good to merge soon

<img src="https://user-images.githubusercontent.com/1945542/202664075-bb921354-6666-4974-a73d-7a8b0b840f1b.png" width="300" />
